src/meta/src/stream/stream_graph/schedule.rs

```diff
@@ -44,88 +55,139 @@ enum Fact {
         to: Id,
         dt: DispatcherType,
     },
-    /// A distribution requirement for a building fragment.
-    InternalReq { id: Id, dist: DistId },
-    /// A distribution requirement for an existing fragment.
+    /// A distribution requirement for an external(existing) fragment.
     ExternalReq { id: Id, dist: DistId },
-    /// The default distribution for a building fragment.
-    Default(DistId),
+    /// A singleton requirement for a building fragment.
+    /// Note that the physical parallel unit is not determined yet.
+    SingletonReq(Id),
+}
+
+/// Results of all building fragments, as the output of the scheduler.
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+enum Result {
+    /// This fragment is required to be distributed by the given [`DistId`].
+    Required(DistId),
+    /// This fragment is singleton, and should be scheduled to the default parallel unit.
+    DefaultSingleton,
+    /// This fragment is hash-distributed, and should be scheduled by the default hash mapping.
+    DefaultHash,
```

Comment:
Instead of providing the default distribution to the datalog, we make the result more explicit and let the caller apply it.